waku/v2/protocol/filter/waku_filter.go

```diff
@@ -0,0 +1,278 @@
+
+	"github.com/status-im/go-waku/waku/v2/protocol"
+	"github.com/status-im/go-waku/waku/v2/protocol/pb"
+)
+
+var log = logging.Logger("wakufilter")
+
+type (
+	ContentFilterHandler func(msg pb.WakuMessage)
+
+	Filter struct {
+		ContentFilters []*pb.FilterRequest_ContentFilter
+		Handler        ContentFilterHandler
+	}
+	// @TODO MAYBE MORE INFO?
+	Filters map[string]Filter
+
+	Subscriber struct {
+		peer      string
+		requestId string
+		filter    pb.FilterRequest // @TODO MAKE THIS A SEQUENCE AGAIN?
+	}
+
+	MessagePushHandler func(requestId string, msg pb.MessagePush)
+
+	WakuFilter struct {
+		ctx         context.Context
+		h           host.Host
+		subscribers []Subscriber
+		pushHandler MessagePushHandler
+		MsgC        chan *protocol.Envelope
+	}
+)
+
+// NOTE This is just a start, the design of this protocol isn't done yet. It
+// should be direct payload exchange (a la req-resp), not be coupled with the
+// relay protocol.
+
+const WakuFilterCodec = "/vac/waku/filter/2.0.0-beta1"
+
+const WakuFilterProtocolId = libp2pProtocol.ID(WakuFilterCodec)
+
+// Error types (metric label values)
+const (
+	dialFailure      = "dial_failure"
+	decodeRpcFailure = "decode_rpc_failure"
+)
+
+func (filters *Filters) Notify(msg *pb.WakuMessage, requestId string) {
+	for key, filter := range *filters {
+		// We do this because the key for the filter is set to the requestId received from the filter protocol.
+		// This means we do not need to check the content filter explicitly as all MessagePushs already contain
+		// the requestId of the coresponding filter.
+		if requestId != "" && requestId == key {
+			filter.Handler(*msg)
+			continue
+		}
+
+		// TODO: In case of no topics we should either trigger here for all messages,
+		// or we should not allow such filter to exist in the first place.
+		for _, contentFilter := range filter.ContentFilters {
+			if msg.ContentTopic == contentFilter.ContentTopic {
+				filter.Handler(*msg)
+				break
+			}
+		}
+	}
+}
+
+func (wf *WakuFilter) selectPeer() *peer.ID {
+	// @TODO We need to be more stratigic about which peers we dial. Right now we just set one on the service.
+	// Ideally depending on the query and our set  of peers we take a subset of ideal peers.
+	// This will require us to check for various factors such as:
+	//  - which topics they track
+	//  - latency?
+	//  - default store peer?
+
+	// Selects the best peer for a given protocol
+	var peers peer.IDSlice
+	for _, peer := range wf.h.Peerstore().Peers() {
+		protocols, err := wf.h.Peerstore().SupportsProtocols(peer, string(WakuFilterProtocolId))
+		if err != nil {
+			log.Error("error obtaining the protocols supported by peers", err)
+			return nil
+		}
+
+		if len(protocols) > 0 {
+			peers = append(peers, peer)
+		}
+	}
+
+	if len(peers) >= 1 {
+		// TODO: proper heuristic here that compares peer scores and selects "best" one. For now the first peer for the given protocol is returned
+		return &peers[0]
+	}
+
+	return nil
+}
```

Comment:
I'll create an issue for refactoring this function, since it's also used in the other protocols

Reply:
Thanks, ok. I just stole it from somewhere else, heh.